eth-contracts/contracts/TokenLocker.sol

```diff
@@ -105,29 +113,66 @@ contract TokenLockerRaw {
-        require(bridgeAmount > bridgeFee, "fee should not exceed bridge amount");
-        uint256 receivedAmount = bridgeAmount - bridgeFee;
-
-        // if token == ETH
-        if (tokenAddress == address(0)) {
-            payable(recipientAddress).transfer(receivedAmount);
-            payable(msg.sender).transfer(bridgeFee);
-        } else {
-            IERC20(tokenAddress).safeTransfer(recipientAddress, receivedAmount);
-            IERC20(tokenAddress).safeTransfer(msg.sender, bridgeFee);
+    function unlockToken(bytes memory proof) external {
+        // 1. getHistoryTxRootInfo from CkbChain
+        (uint64 initBlockNumber, uint64 latestBlockNumber, bytes32 targetHistoryTxRoot) = ckbSpv_.getHistoryTxRootInfo();
+        require(latestBlockNumber > 0, "ckbSpv should initialize");
+
+        // 2. check proveTxRootExist
+        bytes29 txProofVecView;
+        TreeNode[] memory leafNodes;
+        {
+            bytes29 proofView = proof.ref(uint40(ViewSpv.SpvTypes.CKBUnlockTokenParam));
+            bytes29 txRootProofView = proofView.historyTxRootProof();
+            txProofVecView = proofView.historyTxProofVec();
+
+            require(txRootProofView.initBlockNumber() == initBlockNumber, "initBlockNumber mismatch");
+            require(txRootProofView.latestBlockNumber() == latestBlockNumber, "latestBlockNumber mismatch");
+            leafNodes = _proveTxRootExist(txRootProofView, targetHistoryTxRoot);
         }
 
-        emit Unlocked(tokenAddress, recipientAddress, msg.sender, receivedAmount, bridgeFee);
+        // 3. check txs exist and unlockToken
+        bytes29 txProofView;
+        uint64 blockNumber;
+        uint length = txProofVecView.txProofLength();
+        uint64 merkleIndex;
+        for (uint i = 0; i < length; i++) {
+            txProofView = txProofVecView.getHistoryTxProofFromVec(i);
+            blockNumber = txProofView.txBlockNumber();
+            require(
+                blockNumber >= initBlockNumber,
+                "the blockNumber which the tx exists in should be greater than or equal to initBlockNumber"
+            );
+            require(
+                uint256(blockNumber) + uint256(numConfirmations_) <= uint256(latestBlockNumber),
+                "blockNumber from txProofData should not be greater than latestBlockNumber"
+            );
+
+            // - 1. check if txHashes from txProof and raw ckbTx match
+            bytes memory rawTx = txProofView.rawTransaction().clone();
+            bytes32 txHash = CKBCrypto.digest(rawTx, rawTx.length);
+            require(!usedTx_[txHash], "The burn tx cannot be reused");
+            usedTx_[txHash] = true;
+
+            // - 2. proveTxExist, check if txRoots from txProof and txRootProof match
+            // calc the index in txRoot-merkle-tree
+            merkleIndex = (latestBlockNumber - initBlockNumber) + (blockNumber - initBlockNumber);
```

Comment:
Why? Should that be just `blockNumber - initBlockNumber`?

Reply:
- [source code](https://github.com/nervosnetwork/merkle-tree/blob/f615e6a850bb207f47d7003d677e44fa2c92771d/src/merkle_tree.rs#L29
- [doc](https://github.com/nervosnetwork/merkle-tree/blob/master/README.md)
> Suppose a CBMT with n items, the size of the array would be 2n-1, the index of item i(start at 0) is i+n-1. For node at i, the index of its parent is (i-1)/2, the index of its sibling is (i+1)^1-1(^ is xor) and the indexes of its children are [2i+1, 2i+2].

- merkleIndex  =  i+n-1
- n == all_leaves_count == latestBlockNumber - initBlockNumber + 1
- i == index of the item in all_leaves == blockNumber - initBlockNumber
merkleIndex = ( blockNumber - initBlockNumber ) + (latestBlockNumber - initBlockNumber + 1) - 1 = (latestBlockNumber - initBlockNumber) + (blockNumber - initBlockNumber)

Reply:
OK. Can you make it a inline doc for this part? It's not easy to understand.

Reply:
ok